### PR TITLE
Refactor defs to avoid defining non-quantities; expand GWP conversions

### LINKED ISSUE #12: Refactor defs to avoid defining non-quantities
This issue is to carry forward discussion from in https://github.com/IAMconsortium/units/pull/7#issuecomment-597013770, #9, and #10.

## Details

The goal in using pint for unit handling (in the 'client' packages of this repo) is to be _scientific and rigorous_ about measurement.
1. One way pint helps us do this is by ensuring common definitions are used, rather than constants hardcoded in many places across other code.
2. Another way is ensuring precision about _what_ is being measured, and forcing implicit conversions to become explicit.

Per the second point, we should follow, rather than circumvent, the logic of [SI](https://en.wikipedia.org/wiki/International_System_of_Units), as encoded by pint. The [International vocabulary of metrology (VIM)](https://www.bipm.org/utils/common/documents/jcgm/JCGM_200_2008.pdf) is the best reference here, defining e.g.:

**quantity**:
> property  of  a  phenomenon,  body,  or  substance,  where  the  property  has  a  magnitude  that can  be  expressed as a number and a reference.

**quantity value**:
> number and reference together expressing magnitude of a quantity.
>
> EXAMPLE 2 Mass of a given body: 0.152 kg or 152 g
>
> EXAMPLE 8 Mass fraction of cadmium of a given sample of copper: 3 μg/kg or 3 ×10^-9

**nominal property**:
> property of a phenomenon, body, or substance, where the property has no magnitude.
>
> EXAMPLE 1 Sex of a human being.
> EXAMPLE 2 Colour of a paint sample.
> …
> EXAMPLE 5 Sequence of amino acids in a polypeptide.
>
> NOTE 1 A  nominal  property  has  a  value,  which  can  be  expressed  in  words,  by alphanumerical  codes,  or  by  other means. 

In "quantity value" example 8, note that *nominal properties and other information* about the quantity value are supplied *together with* the quantity value:
- The "phenomenon, body, or substance" is "a given sample of copper".
- The quantity value is for the mass fraction of a specific species ("cadmium").

These are essential information. But **only** "µg/kg" is the *unit* (reference) in this quantity value.

As a convention in many disciplines, we often mix these things together, e.g.:
- "vehicle kilometres," "passenger kilometres": in both cases the units are length. Nominal information about the measured phenomenon is "travelled by a {vehicle,person}," sometimes "in a certain period of time."
- GWP (per links above): the unit is mass. The nominal information is that it is the mass of CO₂ that causes global warming equivalent to a given mass of another species.

**tl;dr:**
- Nominal properties are distinct from units.
- Pint is designed to deal with units.
- Forcing pint to handle nominal properties is always going to be awkward.
- We should try to avoid this.

## Actions

- As far as possible, implement definitions and conversions using pint features and *only* physical quantities. 
- Provide example code (e.g. in tests) for how to use these to do common conversions between the *notional* units (i.e. actual units *plus* nominal properties) that we commonly use.
- For each of the categories of quantities in this repo, provide clear description that helps users understand metrology better.

### PR DESCRIPTION
**Pull request by khaeru**
Our first functions! This will close #12 and #11.

The docstrings, README, and new tests show the supported functionality:
https://github.com/IAMconsortium/units/blob/e04629a0c7e38ba1923cfef7e4287b09f473ec30/iam_units/__init__.py#L17-L37
https://github.com/IAMconsortium/units/blob/e04629a0c7e38ba1923cfef7e4287b09f473ec30/iam_units/test_all.py#L97-L119

### CODE REVIEW

File: iam_units/emissions.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    'Halon2402',
+    'N2O',
+    'NF3',
+    'PFPMIE',
+    'SF5CF3',
+    'SF6',
+    'cC3F6',
+    'cC4F8',
+    ]
+
```
Comment (danielhuppmann) [inline iam_units/emissions.py:93]: maybe a little explanation what that one does and where it is needed?

File: iam_units/test_all.py [hunk 2]
```diff
@@ -72,3 +79,41 @@ def test_units_emissions(context, value):
...
+
+
+@pytest.mark.parametrize('units', ['t {}', 'Mt {}', 'Mt {} / yr'])
+@pytest.mark.parametrize('metric, expected_value', EMI_DATA)
+def test_convert_gwp(units, metric, expected_value):
+    # Bare masses can be converted
+
+    qty = registry.Quantity(1.0, units.format(''))
+    expected = registry(f'{expected_value} {units}')
+    assert convert_gwp(metric, qty, 'CH4', 'CO2') == expected
```
Comment (danielhuppmann) [inline iam_units/test_all.py:104]: can this be changed such that just a value is also accepted?

**Comment by danielhuppmann** [comment]
Thanks @khaeru - a few questions as I try to wrap my head around this...

1) I guess having a specific function and the existing context-implementation in parallel is extra hassle and scope for confusion, so do you intend to deprecate the context-gwp-conversion?

2) If yes - does that mean that any tool using this registry will need to implement a switch in its implementation (if emissions/gwp, use `gwp_convert` else do the standard `Qantitity().to()`)?

   This would obviously complicate things in applications relative to current approach... See pyam implementation for reference [here](https://github.com/IAMconsortium/pyam/blob/1170b007e134b811a8d3e9cb173417486ee651c6/pyam/units.py#L34), which is agnostic and simply passes on the context arg if provided - and let's pint figure out what to do.

3) The new implementation is not able to deal with synonyms (which is a really neat feature of pint), not even "CO2e" which is the actual unit after GWP-conversion, not CO2. Is that on the to-do list?

**Comment by khaeru** [comment]
> 1. I guess having a specific function and the existing context-implementation in parallel is extra hassle and scope for confusion, so do you intend to deprecate the context-gwp-conversion?
> 2. If yes - does that mean that any tool using this registry will need to implement a switch in its implementation (if emissions/gwp, use `gwp_convert` else do the standard `Qantitity().to()`)?

I'm still not sure, but was going to ask for such a pointer to the current usage in pyam so this could be adjusted to make it usable. So thanks! If I understand correctly, the user **must** supply the *context* arg if *factor* is None and they wish to do a conversion using a GWP metric—is that right?
```python
        args = [_reg[to]] if context is None else [_reg[to], context]
```
Or put another way, if they are converting emissions species using GWPs and fail to supply *context*, this will not work?

> 3. The new implementation is not able to deal with synonyms (which is a really neat feature of pint), not even "CO2e" which is the actual unit after GWP-conversion, not CO2. Is that on the to-do list?

Yes, I think it could be as simple as adding `a_CO2e = 1.0` here: https://github.com/IAMconsortium/units/blob/e04629a0c7e38ba1923cfef7e4287b09f473ec30/iam_units/data/emissions/emissions.txt#L35-L38

…but, will test. I also need to add `C` for Carbon.

File: iam_units/emissions.py (hunk 1 above)
Comment (khaeru) [reply iam_units/emissions.py:93]: This mainly exists to create the regex on the next line, which is used in convert_gwp. But I can add comments in the generated file.

File: iam_units/test_all.py (hunk 2 above)
Comment (khaeru) [reply iam_units/test_all.py:104]: You mean, just a magnitude? What would its units be?

**Comment by danielhuppmann** [comment]
> Or put another way, if they are converting emissions species using GWPs and fail to supply _context_, this will not work?

Correct, converting a species in pyam to CO2e without context raises a `pint.DimensionalityError`. This is tested [here](https://github.com/IAMconsortium/pyam/blob/1170b007e134b811a8d3e9cb173417486ee651c6/tests/test_units.py#L81).
 
> I also need to add `C` for Carbon.

Yes, please!

File: iam_units/test_all.py (hunk 2 above)
Comment (danielhuppmann) [reply iam_units/test_all.py:104]: Well, in my thinking they would be part of the `*species` args, but I realize now that this would be inconsistent with your goal to get the species out of the units entirely...

**Comment by danielhuppmann** [comment]
Taking a thought from an inline comment to the larger discussion: how would an application like message_ix or pyam keep track of the species when it's not part of the unit...? What is the intended implementation there?

**Comment by khaeru** [comment]
> how would an application like message_ix or pyam keep track of the species when it's not part of the unit

In message_ix reporting, at least, any quantity relating to emissions has an `emission` or `e` dimension that contains the species name. So supposing a pd.DataFrame `all_data` with 'value', 'emission', and 'unit' columns:
```python
for (species, unit), data in all_data.groupby(['emission', 'unit']):
    # Use vector conversion for the entire column
    result = convert_gwp('AR5GWP100',
                         (data['value'].values, unit),
                         species, 'CO2e')
    # Produce whatever kind of output is desired
```
This groups all_data in the largest possible chunks for application of common factors.

**Comment by khaeru** [comment]
CO2e and C are now added.

**Comment by khaeru** [comment]
> I'm still not sure, but was going to ask for such a pointer to the current usage in pyam so this could be adjusted to make it usable.

Also per https://github.com/IAMconsortium/units/pull/16#issuecomment-609501698, here are [all the occurences of 'units' in scmdata](https://github.com/openscm/scmdata/search?p=1&q=units&unscoped_q=units) for reference.

**Comment by danielhuppmann** [comment]
> In message_ix reporting, at least, any quantity relating to emissions has an emission or e dimension that contains the species name.

Pretty sure that this will not be sufficient... If you only record the species and the unit (just the physical quantity), it is not possible to track what the reference is. For example, for carbon dioxide emissions, there is no consensus whether to have the quantity measured in CO2 or C. Similar aspects apply to the HFC species (which are all accounted converted to HFC23-equivalent).

**Comment by khaeru** [comment]
> If you only record the species and the unit (just the physical quantity), it is not possible to track what the reference is. For example, for carbon dioxide emissions, there is no consensus whether to have the quantity measured in CO2 or C. Similar aspects apply to the HFC species (which are all accounted converted to HFC23-equivalent).

You're right, and I'm realizing (and adding this to the README now) there are actually as many as *three* properties:
1. Original emissions species,
2. Species in which the GWP-equivalent are expressed (sometimes CO2, C, or HFC23, as mentioned), and
3. The metric used for conversion.

In various usage that I've seen (could be others):
- (1) is included in a 'variable' column (along with other things) *and* in a 'unit' column (along with the units) when no GWP is applied; (2) and (3) are not applicable.
- If converted, (1) is in a 'variable' column; a 'unit' column contains (2) (instead of (1) as above).
  - Sometimes (3) is explicit, e.g. somewhere in a 'variable' column along with (1).
  - Sometimes (3) is omitted/implicit.
- In the MESSAGE reporting we have:
  - (1) as an `e` dimension,
  - (3) as a `gwp` dimension (see iiasa/message_data#90),
  - We'll probably end up handling (2) by requiring that it always the same internally; or by adding it as another dimension with a constant value.
  - We convert to the above format(s) in order to pass data to existing tools.

This package shouldn't try to accommodate all current/possible ways of arranging that information in downstream code—only to make sure the conversion is done correctly.

**Comment by khaeru** [comment]
@danielhuppmann thanks for these comments!

To explain further, the **motivation** here is the same worries about users:
> Users think of "units" as a combination of a physical and a nominal property.
> 
> There is a risk that users will not follow [the] advice that […] to avoid ambiguity, code handling GHG quantities should also track and output these nominal properties.

In particular, in the iTEM community there have been repeated, circular discussions triggered by confusion over definitions and measurement. The root cause of these was the choice to adopt a data format that is a variant of the IAMC format, and to shoehorn many, conceptually-different things into 'variable' and 'unit' strings in an inconsistent way. Analysis codes are thus very long and convoluted. This is the same cause underlying the need for extensive 'variable'-name mappings to move data between different databases like AR5, ADVANCE, SR15, AR6, etc., when the things measured are often the same.

These experiences are why I see promise in applying a carefully-designed data model like SDMX (or at least its logic), in which concepts are clearly separated. Then, for instance:
- A concept 'GWP metric' can be clearly defined, once.
- In particular applications, according to need:
  - an *attribute* for the metric concept can be attached to an entire data set, or
  - a dataset can have a *dimension* for the metric concept, so it can contain observations with the same quantity converted into CO₂-equivalents by different metrics.

Similarly, in iTEM, we will use concepts (via dimensions and attributes) as far as possible to capture whether `1000 km / year` (magnitude, unit) are travelled by a person, vehicle, or even tonne cargo (different values for a concept of 'thing transported'). (per point 6, there are no standard metrics for e.g. occupancy—`km / year` of persons divided by `km / year` of a vehicle—so that will not go into this repo/package. I may, as mentioned in Slack, remove the base units if I can't convince myself they fit the VIM.)

The *chief goal* of this overarching direction is to clarify the meaning of data and prevent downstream confusion and associated costs/headaches. As an incidental benefit, when these non-unit properties are handled explicitly, units can just be units.

----

As for **implementation**, I think of at least three kinds of code:

1. Low-level utility libraries, like this one.
2. Scientific packages like ixmp and pyam.
3. Researchers' project- or paper-specific analysis codes.

It's good and appropriate for ixmp, pyam, etc. (2) to do some hand-holding for their users (3), depending on their level of knowledge. That can take different forms:
- Silently handle their customary behaviour (without objecting that it's 'impure')—pyam does this, and still will after IAMconsortium/pyam#361.
- Encourage/educate them on how to be more careful and rigorous. We do this often e.g.:
  - by forcing pyam users to pick a specific GWP metric instead of using one by default, we create a small opportunity for them to think, "Which one do I want to use? Which is appropriate for my analysis?"
  - by requiring ixmp users to enter a 'unit' column with parameter data and pre-add units to a Platform, we create opportunities for them to think about units in the first place. (Without the prompt, they might not.)

Low-level libraries (1) can be more opinionated. In the advice "code handling GHG quantities should…" the word "code" mainly applies to (2). IAMconsortium/pyam#361 does this, and the MESSAGE stack is moving in the same direction. But these packages can/should make their own decisions about how forgiving to be of *their* users, per the above.

"code" using this package might also be (3). e.g. if someone uses this package in their research code directly (not via ixmp or pyam). Then they are (choose to be) encouraged to think of origin species, metric, and equivalent species, and how to track them.

----

I guess the tl;dr of the response is, "discipline is useful in downstream packages; we can still choose to be forgiving in user code."

**Comment by danielhuppmann** [comment]
We agree on the key principles:

1) it should be possible to use functions from the low-level utility directly in any code
2) user should be " encouraged to think of origin species, metric, and equivalent species, and how to track them."
3) too many concepts are shoehorned into a variable-units convention in the IAMC framework in a non-harmonised manner

We have different views when it comes to implementation.

The **current implementation** allows the following:
```
registry('1.23 Mt CH4').to('Gt CO2e', 'gwp_AR4GWP100')
```
I think that this obviously satisfies (1) and (2), and there are many things related to units that can be confusing, but methane emissions expressed as "unit" `Gt CO2e` is really not one them - so I think that (3) is also satisfied (in the sense of not over-shoehorning).

This implementation also automatically returns an error when trying to convert emissions without providing a context, so safeguarding against wrong use.

The only "downside" is the violation (I'd call it flexible interpretation) that species are defined as base units.

The **new implementation** defines a new function
```
convert_gwp('AR4GWP100', '1.23 Mt', 'CH4', 'CO2e')
```
which is really not more or less intuitive than the previous one and does not satisfy the three items above better. The only "improvement" is the purity of not defining species as a base unit - but it comes at the expense of substantially more elaborate code in this (what used to be low-level) utility package and any downstream packages.

**tl;dr**: there is a trade-off between discipline (adherence to high-level principles) and simplicity. I'm still convinced that the design decision in this PR unnecessarily overcomplicates this utility and hinders maintainability of the code and any downstream work going forward.

**Comment by khaeru** [comment]
> **tl;dr**: there is a trade-off between discipline (adherence to high-level principles) and simplicity. I'm still convinced that the design decision in this PR unnecessarily overcomplicates this utility and hinders maintainability of the code and any downstream work going forward.

I appreciate this concern. I've also taken the time to read the [code in openscm-units](https://github.com/openscm/openscm-units/blob/master/src/openscm_units/unit_registry.py), and it seems to me this design achieves a superset of features with less code and complexity, and a simpler UX.

As you say at https://github.com/IAMconsortium/units/pull/19#pullrequestreview-390127514, I do take responsibility for these additions. If the functions need adjustment and the code/comments do not, *per se*, show a clear way to write PRs, I commit to making the changes (please open issues), or to explain the code to those who would make them.

As for code downstream, IAMconsortium/pyam#361 preserves and expands current pyam functionality; and using these functions in iiasa/message_data#116 has been straightforward. I'll be happy to provide advice (on short notice) for how best to use this code in pyam, and certainly to expand the test suite here to ensure there is a clearly-defined and stable API for pyam and other client packages to run against.
